protocol/x/listing/keeper/listing.go

```diff
@@ -94,21 +96,29 @@ func (k Keeper) CreateClobPair(
 
 // Function to wrap the creation of a new perpetual
 // Note: This will only list long-tail/isolated markets
-// TODO: Complete implementation and add tests pending marketmap decoding functions and testutils
+// TODO: Add tests pending marketmap testutils
 func (k Keeper) CreatePerpetual(
 	ctx sdk.Context,
 	marketId uint32,
 	ticker string,
 ) (perpetualId uint32, err error) {
 	perpetualId = k.PerpetualsKeeper.AcquireNextPerpetualID(ctx)
 
-	// TODO: Calculate atomic resolution from market map reference price
-	// TODO: get reference price once market map decoding functions are available
-	var atomicResolution int32
-	//marketMapDetails, err := k.MarketMapKeeper.GetMarket(ctx, ticker)
-	//if err != nil {
-	//	return 0, err
-	//}
+	// Get reference price from market map
+	marketMapDetails, err := k.MarketMapKeeper.GetMarket(ctx, ticker)
+	if err != nil {
+		return 0, err
+	}
+	metadata, err := tickermetadata.DyDxFromJSONString(marketMapDetails.Ticker.Metadata_JSON)
+	if err != nil {
+		return 0, err
+	}
+	if metadata.ReferencePrice == 0 {
+		return 0, types.ErrReferencePriceZero
+	}
+
+	// calculate atomic resolution from reference price
+	atomicResolution := -6 - int32(math.Floor(math.Log10(float64(metadata.ReferencePrice))))
```

Comment:
nit: i still think it'd be helpful to create a constant for -6 to make it easy to understand what it represents

Reply:
AFAIK, this was just an arbitrary number to make the resolution closest to existing customs for assets like USDC or BTC. I can create a constant for it